test/integration/cors_filter_integration_test.cc

```diff
@@ -0,0 +1,177 @@
+#include "test/integration/integration.h"
+#include "test/mocks/http/mocks.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+class CorsFilterIntegrationTest : public BaseIntegrationTest,
+                                  public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  CorsFilterIntegrationTest() : BaseIntegrationTest(GetParam()) {}
+  /**
+   * Global initializer for all integration tests.
+   */
+  void SetUp() override {
+    fake_upstreams_.emplace_back(new FakeUpstream(0, FakeHttpConnection::Type::HTTP1, version_));
+    registerPort("upstream_0", fake_upstreams_.back()->localAddress()->ip()->port());
+    createTestServer("test/config/integration/server_cors_filter.json", {"http"});
+  }
+
+  /**
+   * Global destructor for all integration tests.
+   */
+  void TearDown() override {
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+protected:
+  void testPreflight(Http::TestHeaderMapImpl&& request_headers,
+                     Http::TestHeaderMapImpl&& response_headers) {
+    executeActions({
+        // [&]() -> void { response_.reset(new IntegrationStreamDecoder(*dispatcher_)); },
+        [&]() -> void {
+          codec_client_ = makeHttpConnection(lookupPort("http"), Http::CodecClient::Type::HTTP1);
+        },
+        [&]() -> void { codec_client_->makeHeaderOnlyRequest(request_headers, *response_); },
+        [&]() -> void { response_->waitForEndStream(); },
+        [&]() -> void { cleanupUpstreamAndDownstream(); },
+    });
+
+    EXPECT_TRUE(response_->complete());
+    response_headers.iterate(
+        [](const Http::HeaderEntry& entry, void* context) -> void {
+          IntegrationStreamDecoder* response = static_cast<IntegrationStreamDecoder*>(context);
+          Http::LowerCaseString lower_key{entry.key().c_str()};
+          std::string header_value = "";
+          if (response->headers().get(lower_key) != nullptr) {
+            header_value = response->headers().get(lower_key)->value().c_str();
+          }
+          EXPECT_STREQ(entry.value().c_str(), header_value.c_str());
+        },
+        response_.get());
```

Comment:
Sanity check: Are there there non-deterministic headers which make using EXPECT_EQ(response->headers(), expected_response_headers) implausible?   If not I'd suggest adding in whatever extra headers Envoy is and just using EXPECT_EQ.    If there are, please make this a  helper function since it's used in two places.   You can stick at the top of the file, or add a isSubsetOf function to TestHeaderMapImpl since it seems like it might be generically useful.

Reply:
I didn't know expect_eq will work automatically like this. Good to know. But another thing is I want to test the absence of headers. allow-credentials shouldn't be set in many cases for example. Is there a better way to test absence? I had to do the nullptr check or else it would crash of course.

Reply:
If you can craft exact headers, it implicitly tests absence since
{"foo", "bar"} {"eep", "baz"} won't match {"foo", "bar"}.  So if your expected_headers don't have allow-credential and the returned headers did, they'd fail equivalence checks.   The main problem is things like date header where you might have to clear from a cloned reqonse->headers() since it could make your tests flake.

Reply:
You are absolutely right! And it's so obvious! Sorry, didn't think much about the expect_eq. Will amend when I get to work.